### PR TITLE
Request Validation

### LINKED ISSUE #12: Have API Contract and Request Validation with proper error handling
One important aspect of any system is that it should have a well defined and limited universe of inputs that the system will take as input. If the user sends something outside of the universe, system should throw back an appropriate error message. 

We should also have a well defined API contract, and some automated validation while parsing requests that throw appropriate HTTP 400 error.

For example, for CreateQueue API, we should define something like

```
type CreateQueueRequest struct {
    QueueName string `minLength:"4" maxLength:"16"`
}

type CreateQueueResponse model.Queue
```

If an invalid JSON body is received, we should return a 400 Bad Request and the body should indicate to the client the error, `expected , at line 4`. Or if the queueName was supposed to be string but it was not present, or an integer value was passed, that should be pointed out to the user.

### PR DESCRIPTION
**Pull request by aneeshsharma**
As discussed earlier in #12, divided the models among API and DB models
- Added API models in `internal/models/api/`
- Moved DB models to `internal/models/db/`

Implemented some basic validation for a Create Queue request
- Implemented `Validate` method on `CreateQueueRequest`
- Right now it just checks the length of `QueueName` to make sure its between 4 and 20
- Return a 400 HTTP error when the length is outside this range
- Also return 400 HTTP error when unable to parse the JSON

Next steps would be to
- [x] Move Parsing of JSON to middleware
- [x] Move Validation to a middleware

I need to look into how to proceed with implementing middleware in such a way as to have the parsed JSON passed down to subsequent middlewares or the handler.

### CODE REVIEW

File: internal/models/api/queue.go
Comment (maaverik) [inline internal/models/api/queue.go]: nit: Queue Name -> Queue name length

File: internal/handler/queue.go
Comment (maaverik) [inline internal/handler/queue.go]: Can you please rename q to something more expressive like request?

File: internal/models/api/common.go
Comment (daltonfury42) [inline internal/models/api/common.go:2]: Let's add package level comments,

one line comments for each `type`s, and for each field within the type.

File: internal/models/api/common.go
Comment (daltonfury42) [inline internal/models/api/common.go:17]: How about naming this interface as a verb, `Validatable`?

File: internal/models/api/common.go
Comment (daltonfury42) [inline internal/models/api/common.go]: Assuming we will always fail fast at the first validation error, lets keep this as `string` instead of `[]string`?

File: internal/models/api/common.go
Comment (maaverik) [reply internal/models/api/common.go:17]: I think the convention in Go is to name interfaces with an -er suffix, unlike Java (https://go.dev/doc/effective_go#interface-names) so Validator should be fine.

Comment (daltonfury42) [reply internal/models/api/common.go:17]: Another instance of golang being better than java at something. :)

File: internal/models/api/common.go
Comment (aneeshsharma) [reply internal/models/api/common.go]: Ok will make the change

File: internal/models/api/common.go
Comment (aneeshsharma) [reply internal/models/api/common.go:2]: Ok
Should I add for older parts as well?

Comment (daltonfury42) [reply internal/models/api/common.go:2]: We can gradually add for older parts as we touch them.

Comment (aneeshsharma) [reply internal/models/api/common.go:2]: I've added some documentation
I just found this blog post that defines how to do go documentation - https://go.dev/blog/godoc
Seems like there is no docstring constructs and we just need to write comments on top of types, functions etc. No conventions given either.

Please take a look

Comment (maaverik) [reply internal/models/api/common.go:2]: [This](https://elliotchance.medium.com/godoc-tips-tricks-cda6571549b) might help understand conventions @aneeshsharma

Comment (aneeshsharma) [reply internal/models/api/common.go:2]: Thanks @maaverik 
I need to add full stops to paragraphs
I think right now they will be recognized as headings

**Comment by aneeshsharma** [comment]
I started working on implementing a middleware to decode the JSON into our model. But kind of hit a roadblock. I don't think it is possible to implement a single function that can decode all requests into corresponding models since golang doesn't support generics.

Since we cannot decode the request into the API models, it is not possible to use the Validate function as it is right now.

I was thinking that we can instead use `map` objects as requests instead of models for all requests. Then inside the handler we can just extract data from that map. The validator can just check for the kind of request and validate according to that.

The other way would be to decode and validate inside each handler itself. Although I don't think it will be useful since we will have to implement a separate `Validator` for each request anyways.

I would say the `map` approach would make things more simple in all other areas except validation.

What do you think @daltonfury42 @maaverik ?

**Comment by daltonfury42** [comment]
I see. Can I get on a call with you and then we can try the middleware approach together?

I see some support of generics, we can see if it will work: https://go.dev/doc/tutorial/generics

**Comment by aneeshsharma** [comment]
As discussed with @daltonfury42 over call, we should go for decoding and validation of request body inside the handler itself as other approaches are unnecessarily complex specially given the small number of handlers.

**Review by maaverik — DISMISSED** [review]
Could you resolve the file conflicts pls?

**Comment by aneeshsharma** [comment]
Resolved conflicts with `main`

**Review by maaverik — DISMISSED** [review]
Thanks for the PR!

**Comment by maaverik** [comment]
@daltonfury42 any idea why the status checks aren't running here?

**Comment by aneeshsharma** [comment]
I was thinking it might be waiting for the 2 reviews to be completed before running checks.

**Comment by maaverik** [comment]
I'm pretty sure that's not the issue, the checks normally run as soon as code is committed to the PR. I thought it might be due to a setting I had added to make sure the code is up-to-date with main on each PR, but I tried disabling that and reopened this PR (that shouldn't have caused a problem anyway since you merged with main), still nothing. I'll spend time later and try to figure it out, else we'll skip the checks this time.

**Review by daltonfury42 — DISMISSED** [review]

**Comment by daltonfury42** [comment]
My guess is that it will not run against these commits from other action. It's either a bug or some feature to stop infinite loop of commits or maybe a security feature for some sake. 
image


@aneeshsharma can you try adding some empty/random commit to unblock.


Or maybe the commits by an action doesn't qualify as a push.

**Comment by aneeshsharma** [comment]
The checks work now. @daltonfury42 Can you please review the changes?

**Review by daltonfury42 — APPROVED** [review]

**Pull request merged**
